clientapi/routing/room_tagging.go

```diff
@@ -0,0 +1,246 @@
+		dataByte, err := json.Marshal(data)
+		if err != nil {
+			return httputil.LogThenError(req, err)
+		}
+		if err = json.Unmarshal(dataByte, &Tag); err != nil {
+			return httputil.LogThenError(req, err)
+		}
+	}
+	Tag.Tags[tag] = properties
+	err = saveTagData(req, localpart, roomID, accountDB, Tag)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: struct{}{},
+	}
+}
+
+// DeleteTag implements DELETE /_matrix/client/r0/user/{userID}/rooms/{roomID}/tags/{tag}
+func DeleteTag(
+	req *http.Request,
+	accountDB *accounts.Database,
+	device *authtypes.Device,
+	userID string,
+	roomID string,
+	tag string,
+) util.JSONResponse {
+
+	if device.UserID != userID {
+		return util.JSONResponse{
+			Code: http.StatusForbidden,
+			JSON: jsonerror.Forbidden("Cannot set another user's typing state"),
+		}
+	}
+
+	localpart, data, err := obtainSavedTags(req, userID, roomID, accountDB)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+	Tag := newTag()
+
+	// If there are no tags in the database, exit.
+	if len(data) == 0 {
+		//Synapse returns a 200 OK response on finding no Tags, same policy is followed here.
+		return util.JSONResponse{
+			Code: http.StatusOK,
+			JSON: struct{}{},
+		}
+	}
+
+	dataByte, err := json.Marshal(data)
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+	if err := json.Unmarshal(dataByte, &Tag); err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	// Check whether the Tag to be deleted exists
+	if _, ok := Tag.Tags[tag]; ok {
+		delete(Tag.Tags, tag)
+	} else {
+		//Synapse returns a 200 OK response on finding no Tags, same policy is followed here.
+		return util.JSONResponse{
+			Code: http.StatusOK,
+			JSON: struct{}{},
+		}
+	}
+	err = saveTagData(req, localpart, roomID, accountDB, Tag)
+
+	if err != nil {
+		return httputil.LogThenError(req, err)
+	}
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: struct{}{},
+	}
+}
+
+// obtainSavedTags is a utility function to get all the tags saved in the DB
+func obtainSavedTags(
+	req *http.Request,
+	userID string,
+	roomID string,
+	accountDB *accounts.Database,
+) (string, []gomatrixserverlib.RawJSON, error) {
+	localpart, _, err := gomatrixserverlib.SplitID('@', userID)
+	if err != nil {
+		return "", []gomatrixserverlib.RawJSON{}, err
+	}
+
+	data, err := accountDB.GetAccountDataByType(
+		req.Context(), localpart, roomID, "tag",
+	)
+	if err != nil {
+		return "", []gomatrixserverlib.RawJSON{}, err
+	}
+
+	return localpart, extractEventContents(data), nil
+}
+
+// saveTagData is a utility function to save the tag data into the DB
+func saveTagData(
+	req *http.Request,
+	localpart string,
+	roomID string,
+	accountDB *accounts.Database,
+	Tag gomatrix.TagContent,
+) error {
+	newTagData, err := json.Marshal(Tag)
+	if err != nil {
+		return err
+	}
+	if err = accountDB.SaveAccountData(
+		req.Context(), localpart, roomID, "tag", string(newTagData),
```

Comment:
Well, `delete` and `put`  are still operating on a single tag right? Maybe @anoadragon453 has an opinion?

Reply:
There's no conflict though, because `Put` and `Delete` take a list out from DB, modify the list, and put a list back into DB. BTW I'm talking about the column name in the database table, if I've created any confusion.

Reply:
It was clear.

Reply:
I meant, "tag" kinda signifies a single tag being operated on and not necessarily multiple in the case of `Put` and `Delete` which is not in the case of `Get`. So a plural is not necessary is what I was thinking.

Reply:
Did some more investigation and probably this can only be `m.tag`: https://github.com/matrix-org/dendrite/pull/694#issuecomment-512739115

Reply:
As I said, the tests seem to be still failing. I'll need some help in digging up the problems maybe.